client/cgo/coniks.go

```diff
@@ -0,0 +1,182 @@
+)
+
+// main is required to build a shared library, but does nothing
+func main() {}
+
+//export cgoEd25519SignatureVerify
+func cgoEd25519SignatureVerify(pk unsafe.Pointer, pkSize C.int,
+	message unsafe.Pointer, size C.int,
+	sig unsafe.Pointer, sigSize C.int) C.int {
+	if int(pkSize) < crypto.PublicKeySize {
+		return 0
+	}
+	pkBytes := C.GoBytes(pk, pkSize)
+	messageBytes := C.GoBytes(message, size)
+	sigBytes := C.GoBytes(sig, sigSize)
+	if crypto.Verify(pkBytes, messageBytes, sigBytes) {
+		return 1
+	}
+	return 0
+}
+
+//export cgoVrfVerify
+func cgoVrfVerify(pk unsafe.Pointer, pkSize C.int,
+	m unsafe.Pointer, size C.int,
+	index unsafe.Pointer, indexSize C.int,
+	proof unsafe.Pointer, proofSize C.int) C.int {
+	if int(pkSize) < vrf.PublicKeySize ||
+		int(indexSize) < vrf.Size ||
+		int(proofSize) < vrf.ProofSize {
+		return 0
+	}
+	pkBytes := C.GoBytes(pk, pkSize)
+	mBytes := C.GoBytes(m, size)
+	vrfBytes := C.GoBytes(index, indexSize)
+	proofBytes := C.GoBytes(proof, proofSize)
+	if vrf.Verify(pkBytes, mBytes, vrfBytes, proofBytes) {
+		return 1
+	}
+	return 0
+}
+
+//export cgoHashChainVerify
+func cgoHashChainVerify(prevHash unsafe.Pointer, root unsafe.Pointer, size C.int) C.int {
+	if int(size) != crypto.HashSizeByte {
+		return 0
+	}
+	prevHashBytes := C.GoBytes(prevHash, size)
+	rootBytes := C.GoBytes(root, size)
+	if bytes.Equal(prevHashBytes, rootBytes) {
+		return 1
+	}
+	return 0
+}
+
+//export cgoAuthenticationPathVerify
+//must call cgoVrfVerify first to verify the returned vrf index.
+func cgoAuthenticationPathVerify(treeHash unsafe.Pointer, treeHashSize C.int,
+	treeNonce unsafe.Pointer, treeNonceSize C.int,
+	lookupIndex unsafe.Pointer, indexSize C.int,
+	prunedHashes unsafe.Pointer, prunedSize C.int, hashSize C.int,
+	leafLevel C.int,
+	leafIndex unsafe.Pointer, leafIndexSize C.int,
+	leafCommitment unsafe.Pointer, commitmentSize C.int,
+	isLeafEmpty C.int) C.int {
+
+	if int(treeHashSize) != crypto.HashSizeByte ||
+		int(treeNonceSize) != crypto.HashSizeByte ||
+		int(indexSize) != vrf.Size ||
+		int(hashSize) != crypto.HashSizeByte ||
+		int(prunedSize) < 1 {
+		return 0
+	}
+	if int(isLeafEmpty) != 1 && // user leaf node
+		(int(leafIndexSize) != vrf.Size ||
+			int(commitmentSize) != crypto.HashSizeByte) {
+		return 0
+	}
+
+	th := C.GoBytes(treeHash, treeHashSize)
+	tn := C.GoBytes(treeNonce, treeNonceSize)
+	li := C.GoBytes(lookupIndex, indexSize)
+	leafi := C.GoBytes(leafIndex, leafIndexSize)
+	leafc := C.GoBytes(leafCommitment, commitmentSize)
+
+	buf := C.GoBytes(prunedHashes, prunedSize*hashSize)
+	pt := make([][]byte, 0, prunedSize)
+	for i := 0; i < int(prunedSize); i++ {
+		pt = append(pt, buf[:hashSize])
+		buf = buf[hashSize:]
+	}
+
+	ap := new(merkletree.AuthenticationPath)
+	ap.TreeNonce = tn
+	ap.LookupIndex = li
+	ap.PrunedTree = pt
+
+	if verifyProof(ap, leafi, leafc, int(leafLevel), int(isLeafEmpty), th) {
+		return 1
+	}
+	return 0
+}
+
+func computeLeafHash(ap *merkletree.AuthenticationPath,
```

Comment:
`computeLeafHash`, `authPathHash`, `verifyProof` are presumably the same from proof_test.go, and should be deduped.  Maybe just expose `merkletree.VerifyProof`?